geomstats/geometry/hamiltonian_geodesics.py

```diff
@@ -0,0 +1,241 @@
+        tangent_vec = self.make_symmetric(gs.reshape(res.x, base_point.shape))
+        return tangent_vec
+
+    def scipy_log(self, point, base_point):
+        stop_time = 1.0
+        t = gs.linspace(0, stop_time, self.n_steps + 2)
+        s = time()
+
+        def bvp(time, flat_states):
+            result = []
+            for flat_state in flat_states.T:
+                state_ = flat_state.reshape(gs.stack([base_point, point]).shape)
+                result.append(self.symp_grad(self.kinetic_energy)(state_).flatten())
+            return gs.stack(result).T
+
+        def boundary_cond(flat_state_0, flat_state_1):
+            state_0 = flat_state_0.reshape((2, -1))
+            state_1 = flat_state_1.reshape((2, -1))
+            return gs.array(
+                [state_0[0] - base_point.flatten(), state_1[0] - point.flatten()]
+            ).flatten()
+
+        geodesic_init = gs.einsum("n,kij->nkij", t, point) + gs.einsum(
+            "n,kij->nkij", 1 - t, base_point
+        )
+        velocity_init = gs.stack([gs.zeros_like(base_point)] * (self.n_steps + 2))
+        state_init = gs.stack([geodesic_init, velocity_init], axis=1)
+        state_init = state_init.transpose(1, 2, 3, 4, 0).reshape((-1, self.n_steps + 2))
+        solution = solve_bvp(bvp, boundary_cond, t, state_init, tol=1e-6)
+        geodesic = solution.y
+        geodesic = geodesic[:, :2]
+        velocity = 1.0 / solution.x[1] * (geodesic[:, 1] - geodesic[:, 0])
+        velocity = velocity.reshape((2, -1))[0].reshape(base_point.shape)
+        print(f"Finished path straightening in {time() - s} sec")
+
+        return velocity
+
+    def kinetic_energy(self, state):
+        position, momentum = state
+        momentum = self.make_symmetric(momentum)
+        # position = self.make_symmetric(position)
+        return 1 / 2 * cometric(self.power1, self.power2, momentum, momentum, position)
+
+    @staticmethod
+    def make_symmetric(mat):
+        return 1.0 / 2 * (mat + gs.transpose(mat, axes=(0, 2, 1)))
+
+    @classmethod
+    def symp_flow(cls, hamiltonian, end_time=1.0, n_steps=20, scheme="euler"):
+        if scheme == "rk2":
+            step = cls.rk2
+        elif scheme == "leapfrog":
+            step = cls.leapfrog
+        elif scheme == "rk4":
+            step = cls.rk4
+        elif scheme == "yoshida":
+            step = cls.yoshida
+        else:
+            step = cls.symp_euler
+        step_size = end_time / n_steps
+        return cls.iterate(step(hamiltonian, step_size), n_steps)
+
+    @classmethod
+    def symp_grad(cls, hamiltonian):
+        def vector(x):
+            H_q, H_p = elementwise_grad(hamiltonian)(x)
+            return gs.array([H_p, -H_q])
+
+        return vector
+
+    @classmethod
+    def symp_euler(cls, hamiltonian, step_size):
+        def step(state):
+            position, momentum = state
+            dq, _ = cls.symp_grad(hamiltonian)(state)
+            y = gs.array([position + dq, momentum])
```

Comment:
This line should be:

y = gs.array([position + step_size * dq, momentum])

(that fixed my wrong results when integrating the Hamiltonian eq.'s in the SubRiemannianMetric class)